### PR TITLE
Sot 42 quadratic arithmetic programs from zero to hero

### PR DESCRIPTION
**Pull request by ktokto313**
- [x] Did you check your grammar / reader proof your document using ChatGPT?
- [x] Don't forget to squash commits into meaningful chunks before merging
- [x] After merging this, please update your task status!

This pull request only consist of my commit, the other commits is from SOT-39

### CODE REVIEW

File: articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic programs_from_zero_to_hero.md
Comment (sonntuet1997) [inline articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic programs_from_zero_to_hero.md:1]: filename

File: articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic programs_from_zero_to_hero.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+or 
+
+$x = y$ (op) $z$ (where op can be +, -, *, / and y and z can be variables, numbers or sub-expressions)
+
+You can think of those as logic gates in a circuit.
+
+The result of the flattening process for the above equation is as follows:
+
+$sym_1 = x * x$
```
Comment (haipro287) [inline articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic programs_from_zero_to_hero.md:21]: break line

File: articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic programs_from_zero_to_hero.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+so you might be asking how do we get  $A . s * B . s - C . s = 0$ to be true. So it all come to the property of dot product.
+
+For example:
+
+if you have a vector [1, 2, 3, 4, 5, 6]
+
+then dot product it with vector [0, 0, 1, 0, 0, 0]
+
+we have [0, 0, 3, 0, 0, 0] as the result
```
Comment (haipro287) [inline articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic programs_from_zero_to_hero.md:93]: result of dot product must be a single value

**Comment by ktokto313** [comment]
All 3 previous conversations have been resolved in the latest commit!

File: articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+Turn all columns of A 
+into a matrix of polynomial f(x) of n-1 degree using _Lagrange interpolation_ 
+
+For example: 
+
+the first column of matrix A is 
+
+0 0 0 5
+
+so we will use _Lagrange interpolation_ to find a polynomial of 3 degree which equal to those 4 values at x=1, x=2, x=3 and x=4 or we can say that the polynomial go through 4 points (1, 0), (2, 0), (3, 0), (4, 5)
```
Comment (sonntuet1997) [inline articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md:116]: Link to our definition of Lagrange interpolation

File: articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md
Comment (sonntuet1997) [inline articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md:1]: Please check the grammar and punctuation

File: articles/proofs-arguments-and-zero-knowledge/chapter_2.md [hunk 4]
```diff
@@ -74,9 +74,9 @@ Reason: Time complexity of multiplication of a (n x n) matrix and a vector of si
 
    Consider matrices:
 
-   $`A = \begin{bmatrix} 1 & 2 & 3 \\ 4 & 5 & 6 \\ 7 & 8 & 9 \end{bmatrix}`$
+   $A = \begin{bmatrix} 1 & 2 & 3 \\ 4 & 5 & 6 \\ 7 & 8 & 9 \end{bmatrix}$
```
Comment (sonntuet1997) [inline articles/proofs-arguments-and-zero-knowledge/chapter_2.md:77]: math error?

File: articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+and so on
+
+then we have 
+
+[$f_{A1} (1)$, $f_{A2} (1)$, $f_{A3} (1)$, $f_{A4} (1)$, $f_{A5} (1)$]
+
+equal to the entire first row of A
+so instead of checking R1CS one by one, we now have three set of polynomials which give $A(x).s * B(x).s - C(x).s = 0$ at x =1, x = 2, x = 3, x = 4 
+
+or we say that the equation t(x) = 0 have root on x=[1, 2, 3, 4]
```
Comment (haipro287) [inline articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md:134]: math

File: articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+or t(x) have the form of $(x-1)(x-2)(x-3)(x-4)h(x)$ with h(x)
```
Comment (haipro287) [inline articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md:136]: math

File: articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+for faster check, divide it with a basic polynomial Z which is $(x-1)(x-2)(x-3)(x-4)$
+
+however we can trick this system like use the input where A = C and B with x = 1
```
Comment (haipro287) [inline articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md:140]: I think you should show the final equation of this step here

File: articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md
Comment (ktokto313) [reply articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md:1]: I got GPT do it for me, I think it should be ok now.

File: articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md (hunk 3 above)
Comment (ktokto313) [reply articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md:113]: new term file created

**Comment by sonntuet1997** [comment]
Fix the commit message

**Comment by sonntuet1997** [comment]
SOT-37: Move fundamental_theorem_of_algebra.md Harry Nguyen* 17/12/2023, 15:36 ??

**Comment by sonntuet1997** [comment]
Rebase?

**Comment by ktokto313** [comment]
> SOT-37: Move fundamental_theorem_of_algebra.md Harry Nguyen* 17/12/2023, 15:36 ??

As I have told you before, this branch's origin is from SOT-37 branch, so it has some SOT-37 commit in here

File: articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md (hunk 6 above)
Comment (ktokto313) [reply articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md]: done

File: articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md (hunk 7 above)
Comment (ktokto313) [reply articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md]: ok

File: articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md (hunk 5 above)
Comment (ktokto313) [reply articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md]: ok

File: articles/proofs-arguments-and-zero-knowledge/chapter_2.md (hunk 4 above)
Comment (ktokto313) [reply articles/proofs-arguments-and-zero-knowledge/chapter_2.md:77]: I haven't touched anything about chapter 2, but this doesn't appear to be an error for me either

**Comment by ktokto313** [comment]
All stuff fixed, please review!

File: articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+$[5, 0, 0, 0, 0, 1]$
+
+B  
+$[0, 1, 0, 0, 0, 0]$  
+$[0, 1, 0, 0, 0, 0]$  
+$[1, 0, 0, 0, 0, 0]$  
+$[1, 0, 0, 0, 0, 0]$
+
+C  
+$[0, 0, 0, 1, 0, 0]$
```
Comment (haipro287) [inline articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md:82]: break line

Comment (ktokto313) [reply articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md:91]: resolved

**Comment by sonntuet1997** [comment]
Please rebase

**Review by VanhGer — APPROVED** [review]
côl

File: terms/knowledge_of_exponent.md [hunk 9]
```diff
@@ -0,0 +1,43 @@
+# Knowledge Of Exponent (KOE)
+
+If you get a pair of points $P$ and $Q$, where $P * k = Q$, and you get a point $C$, then it is not possible to come up
+with $C * k$ unless $C$ is “derived” from $P$ in some way that you know.
+
+P = 2
```
Comment (haipro287) [inline terms/knowledge_of_exponent.md:6]: break line and put them in $$

File: terms/knowledge_of_exponent.md [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+C = P * 2 = 4
+D = C *k = P *2 * k = (P * k) * 2 = Q * 2 = 32
+
+C, D ? => P, Q (k)
+C * Q ? D * P
+C * P * k ? C * k * P
+
+Supposed that a pair of points $(P, Q)$ falls from the sky, where $P * k = Q$, but nobody knows what the value of $k$
+is. Now, suppose that I come up with a pair of points $(R, S)$ where $R * k = S$. Then, the knowledge-of-exponent assumption implies that the only way I could have made that pair of points was by taking $P$ and $Q$, and multiplying
```
Comment (haipro287) [inline terms/knowledge_of_exponent.md:18]: I think using "we" instead of "I" will help make the article more objective

Comment (haipro287) [reply terms/knowledge_of_exponent.md:18]: Oh is this paragraph a copy-paste?

File: articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+$f_{A1}(2)$ = first column second row
+
+and so on
+
+then we have
+
+$[f_{A1} (1), f_{A2} (1), f_{A3} (1), f_{A4} (1), f_{A5} (1)]$
+
+equal to the entire first row of A
+so instead of checking R1CS one by one, we now have three set of polynomials
```
Comment (sonntuet1997) [inline articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md:152]: set(s)

File: .obsidian/workspace.json [hunk 12]
```diff
@@ -35,14 +35,26 @@
             "state": {
               "type": "markdown",
               "state": {
-                "file": "articles/proofs-arguments-and-zero-knowledge/chapter_2.md",
+                "file": "articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md",
```
Comment (thuan2172001) [inline .obsidian/workspace.json]: Remove obsidian

File: articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+Rank-1 Constraint System (R1CS) transforms each flattened equation into a constraint system using the formula: $A . s *
+B . s - C . s = 0$, where A, B, C are vectors representing variable mappings, and s is the full variable list.
+
+Transform the equation as follows:
+
+$$
+\begin{align}
+C &= A * B \\
+A * B - C &= 0 \\
```
Comment (thuan2172001) [inline articles/quadratic-arithmetic- programs-from-zero-to-hero/quadratic_arithmetic_programs_from_zero_to_hero.md:38]: Missing EQUIVALENT operation

File: articles/zk-snarks-under-the-hood/zk_snarks_under_the_hood.md [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+actually create the accompanying $Q_1…Q_{10}$ points without knowing what $k$ is, and if you do know what $k$ is then
+you can create a pair $(R, S)$ where $R * k = S$ for whatever $R$ you want, without bothering to create a linear
+combination. ==Hence, for this to work it’s absolutely imperative that whoever creates those points is trustworthy and
+actually deletes k once they created the ten points.== ==**This is where the concept of a “trusted setup” comes from
+**====.==
+
+Remember that the solution to a QAP is a set of polynomials $(A, B, C)$ such that $A(x) * B(x) - C(x) = H(x) * Z(x)$,
+where:
+
+- A is a linear combination of a set of polynomials {$A_1…A_m$}
```
Comment (thuan2172001) [inline articles/zk-snarks-under-the-hood/zk_snarks_under_the_hood.md:40]: Cannot render math expressions

File: articles/zk-snarks-under-the-hood/zk_snarks_under_the_hood.md [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+
+[Pinocchio protocol](https://eprint.iacr.org/2013/279.pdf) Parno, Gentry, Howell and Raykova from 2013 (often called
+PGHR13)
+
+Supposed that a pair of points $(P, Q)$ falls from the sky, where $P * k = Q$, but nobody knows what the value of $k$
+is. Now, suppose that I come up with a pair of points $(R, S)$ where $R * k = S$. Then,
+the [knowledge of exponent](../../terms/knowledge_of_exponent.md) assumption implies that the only way I could have made
+that pair of points was by taking $P$ and $Q$, and multiplying both by some factor r_that I personally know_.
+
+Checking $R * k = S$_doesn’t actually require knowing k -_instead, you can simply check whether or not $e(R, Q) = e(P,
```
Comment (thuan2172001) [inline articles/zk-snarks-under-the-hood/zk_snarks_under_the_hood.md:20]: Check this math expression

Comment (ktokto313) [reply articles/zk-snarks-under-the-hood/zk_snarks_under_the_hood.md:20]: this is not the file related to this Pull request, so we shouldn't care about this, the PR for this is SOT 39

**Review by sonntuet1997 — APPROVED** [review]

**Review by sonntuet1997 — COMMENTED** [review]
Remember to squash

**Review by Tranduy1dol — APPROVED** [review]

**Pull request merged**